// gpstracker/lib/tracker.js
var EventEmitter = require("events").EventEmitter;

function Tracker(client, imei){
  EventEmitter.call(this);

  this.imei = imei;
  this.client = client;
}

Tracker.prototype = Object.create(EventEmitter.prototype);

/*
 * usage: tracker.trackEvery(10).seconds();
 *        tracker.trackEvery(1).hours();
 *        tracker.trackEvery(10).meters();
 */
Tracker.prototype.trackEvery = function(value){
  var result = {},
      thisTracker = this, 
      multiTrackFormat = {
        "seconds": [2,"s"],
        "minutes": [2,"m"],
        "hours":   [2,"h"],
        "meters":  [4,"m"]
      };
  Object.keys(multiTrackFormat)
    .forEach(function(k){
      result[k] = function(){
        var format = multiTrackFormat[k],
            interval = Array(format[0] - String(value).length + 1).join('0')+ value + format[1],
            message = "**,imei:" + thisTracker.imei + ",C," + interval;
        thisTracker.client.write(new Buffer(message));
      };
    });

  return result;
};


//Gets single position
Tracker.prototype.getPosition = function(){
  this.client.write(new Buffer("**,imei:" + this.imei + ",B"));
};

//Position Cancel
Tracker.prototype.positionCancel = function(){
  this.client.write(new Buffer("**,imei:" + this.imei + ",D"));
};

//Cancel Alarm
Tracker.prototype.alarmCancel = function(){
  this.client.write(new Buffer("**,imei:" + this.imei + ",E"));
};


//Move Alarm
Tracker.prototype.moveAlarm = function(){
  this.client.write(new Buffer("**,imei:" + this.imei + ",E"));
};

//Overspeed alarm
//Over speed alarm(speed 60KM/hour,it has 10-300 for option)
Tracker.prototype.overspeedAlarm = function(speed){
  //TODO: format speed as 3 digits (eg 060) 
   this.client.write(new Buffer("**,imei:" + this.imei + ",H,"+speed));
};


module.exports = Tracker;
